ap/inference/datacard.py

```diff
@@ -0,0 +1,550 @@
+# coding: utf-8
+
+"""
+Helpers to write and work with datacards.
+"""
+
+import os
+from collections import OrderedDict
+from typing import Optional, Dict, Tuple, List, Sequence, Any
+
+import law
+
+from ap.inference import InferenceModel, ParameterType, ParameterTransformation
+from ap.util import DotDict, maybe_import, real_path, ensure_dir, safe_div
+
+np = maybe_import("np")
+hist = maybe_import("hist")
+uproot = maybe_import("uproot")
+
+
+logger = law.logger.get_logger(__name__)
+
+
+class DatacardWriter(object):
+    """
+    Generic writer for combine datacards using a instance of an :py:class:`InferenceModel`
+    *inference_model_inst* and a threefold nested dictionary "category -> process -> shift -> hist".
+
+    *rate_precision* and *parameter_precision* control the number of digits of values for measured
+    rates and parameter effects.
+
+    .. note::
+
+        At the moment, all shapes are written into the same root file and a shape line with
+        wildcards for both bin and process resolution is created.
+    """
+
+    def __init__(
+        self,
+        inference_model_inst: InferenceModel,
+        histograms: Dict[str, Dict[str, Dict[str, hist.Hist]]],
+        rate_precision: int = 4,
+        parameter_precision: int = 4,
+        sep: str = "  ",
+    ):
+        super().__init__()
+
+        # store attributes
+        self.inference_model_inst = inference_model_inst
+        self.histograms = histograms
+        self.rate_precision = rate_precision
+        self.parameter_precision = parameter_precision
+        self.sep = sep
+
+    def write(
+        self,
+        datacard_path: str,
+        shapes_path: str,
+        shapes_path_ref: Optional[str] = None,
+    ) -> None:
+        """
+        Writes the datacard into *datacard_path* with shapes saved in *shapes_path*. When the paths
+        exhibit the same directory and *shapes_path_ref* is not set, the shapes file reference is
+        relative in the datacard.
+        """
+        # determine full paths and the shapes path reference to put into the card
+        datacard_path = real_path(datacard_path)
+        shapes_path = real_path(shapes_path)
+        if not shapes_path_ref:
+            shapes_path_ref = os.path.relpath(shapes_path, os.path.dirname(datacard_path))
+            if shapes_path_ref.startswith(".."):
```

Comment:
Mh, then in your datacards, the shape files were referenced as e.g. "../root_files/file1.root"?
This could still be done by setting the `shapes_path_ref` explicitly to this value.